BlackCubes/foodtrucks-django-react-native
# Custom User Model

## Changes
1. Initiated a user app.
2. Created custom User model and manager.
3. Created a basic test for the User model.
4. Configured image uploads along with installing `Pillow` package.

## Purpose
Although Django comes with a User Model out of the bag, a custom model needs to be created for normal and admin users in order to require other fields.

## Approach
The first step is to create a new Django app by calling it `user`. Next, is to create a custom User model from scratch (hence the `AbstractBaseUser` with required `PermissionsMixin` for doing it from scratch) with fields of:
- uuid
- email
- username
- profile_image
- is_staff
- is_active
- date_joined

To discard the default Django fields of `first_name` and `last_name` from the model, they were set to `None`.

The next step is to create a custom manager in order to tell Django on the new fields for the custom User model for both regular users and admin users (works as well when first using the `createsuperuser` on the terminal). Then, in the User model, the custom user manager is put inside the model which tells Django that the model is using a custom user manager.

An `ImageField` class was used for `profile_image` property in the model, and to make image uploads work, a `MEDIA_ROOT` was defined in `settings.py` which tells Django where the user uploaded files are going to be at which in this case is at the global level in the `/server` folder called `/media`. A `Pillow` package was installed in order to make image uploads work since there is an `ImageField` (if using `FileField`, then `Pillow` would not have to be installed). To create unique image names to the particular user and uploaded to a `/user` folder, a `upload_to` function was created and used in `profile_image`.

A basic test was also created for the User model.

Last of all, in order to tell Django to use the custom User model, in `settings.py`, a `AUTH_USER_MODEL` was created with a value of `user.CustomUser`.

## Testing
1. Pull changes.
2. Open the terminal, go to the `/server` folder in your terminal, activate the virtual environment by running `env/Scripts/activate` on Windows, or `env/bin/activate` on Mac.
3. Then, install `Pillow` by running `pip install Pillow`, or from `requirements.txt` by using `pip install -r requirements.txt`.
4. Create a new folder called "media" in the `/server` folder.
5. Type in the terminal `python manage.py makemigrations --dry-run --verbosity 3` and make sure the tests fail.
6. DO NOT MAKE THE MIGRATIONS UNTIL ALL MODELS ARE FINISHED!

Closes #3
